Rework stock_movement_status calculation

jniles opened the pull request:
This is a full rewrite of the `stock_movement_status` table and associated stored procedures.  The goal of this rewrite is to improve clarity, remove user-defined variables that may potentially collide across MySQL connections, and improve the performance of the `GetCMM()` calculation.


Closes #5073.
Closes #5144.
Closes https://github.com/IMA-WorldHealth/bhima/issues/5332.
Closes https://github.com/IMA-WorldHealth/bhima/issues/5360.


----
![image](https://user-images.githubusercontent.com/896472/108206217-6741e600-7126-11eb-86c1-ada342009708.png)


![image](https://user-images.githubusercontent.com/896472/108207993-c1dc4180-7128-11eb-9bf7-c611e20ec500.png)


![heidisql_0CKbDzKdzl](https://user-images.githubusercontent.com/896472/108201428-17602080-7120-11eb-81c5-723e8cc122e9.png)



This PR contains many far reaching changes.  Here is a lot of the notable ones:

 1. I've removed the `avg_consumption` column from the inventory table.  It was never used, as far as I can tell, to do any reporting.  It was used in the stock import feature, but we don't use the `avg_consumption` to report on, so I just removed that code.
 2. I've removed the `stock_consumption` table and associated procedures.  It is used in several reports and they are now broken.  However, I would argue they were always broken since the `stock_consumption` did not record accurate information about stock consumption.
 3.  I've entirely changed the structure behind the `stock_movement_status` table for the following reasons:
     1. The [previous structure](https://github.com/IMA-WorldHealth/bhima/blob/e455a8011db69ad48839afc35977d3dd041d3241/server/models/schema.sql#L2016) was very hard to audit.  For example, it had `start_date`, `end_date`, and `quantity`.  Is that the quantity at the start?  The quantity at the end?  If I wanted to find the value on date X, I needed to do a BETWEEN query.
     2. The table did not distinguish between consumption information and exit information.  This made the `GetCMM()` calculation have to look up additional information from the `stock_movement` table.  A user would need to as well.
     3. In theory, we'll be calling `GetCMM()` more than we will be moving stock.  It makes sense to make the performance tradeoff to increase the speed of `GetCMM()` and decrease the speed of inserting stock.
 5. I've renamed the `GetCMM()` call to `GetAMC()` for clarity.  It also takes in a single date and uses the depot to find the correct value of the date range.  This prevents us from having inconsistent results because a developer didn't properly compute the date range.  If we need to support a custom date range, we'll just write another SP.
 6. I've slightly changed the API of `GetAMC()`:
     1. It returns `head_days` and `tail_days` instead of `days_before_stock_consumption`.  That is because `days_before_stock_consumption` is the real number of days before the first consumption.  Instead, it the days between the start of the window and the next record.  Similarly, `tail_days` gives you the number of days between the last consumption record and the end of the window.
     2. It returns the `quantity_in_stock` so that we can perform stock out date calculations.
     3. `first_inventory_movement_date` -> `min_date`
     4. `last_inventory_movement_date` -> `max_date`

    
------

It isn't all good.  In fact, initializing this is terrible.  On my machine (4GB of RAM):
```mysql
mysql> use vanga;
mysql> CALL zRecomputeStockMovementStatus();
Query OK, 0 rows affected (22 min 4.81 sec)

mysql> use imck;
mysql> CALL zRecomputeStockMovementStatus();
Query OK, 0 rows affected (4 min 20.31 sec)
```

This is _horrible_.  I did get a lot faster speeds when I was just testing on MySQL 8, but the syntax didn't work on MySQL 5.7 (#5364).  When I fixed it on MySQL5.7, I ran into issues on MySQL 8.  Basically, we got the worst performance of both worlds by trying to get it to work correctly on both worlds.

However, I still think it is worth merging because the performance of any individual transaction isn't bad.  It is only when rebuilding the database that things go horribly long.

----

### How to test

Download this PR and run the tests!  (`yarn test:integration` and `yarn test:integration:stock`).  If that works, try with a production database.  Note, as above, you'll need to `CALL zRecomputeStockMovementStatus()` to build the table.

dependabot-preview[bot] committed:
chore(deps-dev): bump karma from 6.0.4 to 6.1.0

Bumps [karma](https://github.com/karma-runner/karma) from 6.0.4 to 6.1.0.
- [Release notes](https://github.com/karma-runner/karma/releases)
- [Changelog](https://github.com/karma-runner/karma/blob/master/CHANGELOG.md)
- [Commits](https://github.com/karma-runner/karma/compare/v6.0.4...v6.1.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps-dev): bump release-it from 14.2.2 to 14.3.0

Bumps [release-it](https://github.com/release-it/release-it) from 14.2.2 to 14.3.0.
- [Release notes](https://github.com/release-it/release-it/releases)
- [Changelog](https://github.com/release-it/release-it/blob/master/CHANGELOG.md)
- [Commits](https://github.com/release-it/release-it/compare/14.2.2...14.3.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

jniles committed:
fix(inventory): remove avg_consumption column

This commit removes the avg_consumption column in the inventory table
since it does not play a role anymore. Additionally, it unifies the
inventory databases to using the same test/data/inventories.sql file.

Closes #5144.

jniles committed:
chore(stock): remove stock_consumption table

Removes the stock_consumption table and changes SQL references to no
longer reference this table.  Passes integration tests, but we still
need to update to stock_movement_status in several reports.

Closes #5073.

jniles committed:
chore(stock): rm currencyId in PostStockMovement

Removes the requirement to pass in the currencyId to the
PostStockMovement() function.  We always use the enterprise currency, so
we can just look this up with the provided projectId.

dependabot-preview[bot] committed:
chore(deps-dev): bump chai from 4.2.0 to 4.3.0

Bumps [chai](https://github.com/chaijs/chai) from 4.2.0 to 4.3.0.
- [Release notes](https://github.com/chaijs/chai/releases)
- [Changelog](https://github.com/chaijs/chai/blob/master/History.md)
- [Commits](https://github.com/chaijs/chai/compare/4.2.0...4.3.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps): bump @ima-worldhealth/coral from 2.4.0 to 2.5.1

Bumps @ima-worldhealth/coral from 2.4.0 to 2.5.1.

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps): bump ioredis from 4.19.4 to 4.22.0

Bumps [ioredis](https://github.com/luin/ioredis) from 4.19.4 to 4.22.0.
- [Release notes](https://github.com/luin/ioredis/releases)
- [Changelog](https://github.com/luin/ioredis/blob/master/Changelog.md)
- [Commits](https://github.com/luin/ioredis/compare/v4.19.4...v4.22.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

jniles committed:
wip(stock): refactor stock_movement_status calc

Work in progress refactoring the calculations for stock_movement_status.

This commit reworks the stock movement status to that the running
balances are kept in the the stock_movement_status table and
computations are done without LOOPs in the stored procedure.

jmcameron committed:
Display inventory name in merge lots modal

dependabot-preview[bot] committed:
chore(deps): bump connect-redis from 5.0.0 to 5.1.0

Bumps [connect-redis](https://github.com/visionmedia/connect-redis) from 5.0.0 to 5.1.0.
- [Release notes](https://github.com/visionmedia/connect-redis/releases)
- [Commits](https://github.com/visionmedia/connect-redis/compare/v5.0.0...v5.1.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

jniles committed:
feat(stock): rewrite stock_movement_status

This commit features a complete rewrite of the stock_movement_status
table and associated stored procedures.  It should be much more
efficient, reliable, and auditable.  It stores the running total in the
table and only invalidates/updates the information necessary.

jmcameron committed:
Add stock entry typeahead support for selecting existing lots

dependabot-preview[bot] committed:
chore(deps-dev): bump typescript from 4.1.3 to 4.1.4

Bumps [typescript](https://github.com/Microsoft/TypeScript) from 4.1.3 to 4.1.4.
- [Release notes](https://github.com/Microsoft/TypeScript/releases)
- [Commits](https://github.com/Microsoft/TypeScript/compare/v4.1.3...v4.1.4)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

mbayopanda committed:
Adding date in the stock import module

dependabot-preview[bot] committed:
chore(deps-dev): bump typescript from 4.1.4 to 4.1.5

Bumps [typescript](https://github.com/Microsoft/TypeScript) from 4.1.4 to 4.1.5.
- [Release notes](https://github.com/Microsoft/TypeScript/releases)
- [Commits](https://github.com/Microsoft/TypeScript/compare/v4.1.4...v4.1.5)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps-dev): bump mocha from 8.2.1 to 8.3.0

Bumps [mocha](https://github.com/mochajs/mocha) from 8.2.1 to 8.3.0.
- [Release notes](https://github.com/mochajs/mocha/releases)
- [Changelog](https://github.com/mochajs/mocha/blob/master/CHANGELOG.md)
- [Commits](https://github.com/mochajs/mocha/compare/v8.2.1...v8.3.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps-dev): bump chai-datetime from 1.7.0 to 1.8.0

Bumps [chai-datetime](https://github.com/mguterl/chai-datetime) from 1.7.0 to 1.8.0.
- [Release notes](https://github.com/mguterl/chai-datetime/releases)
- [Commits](https://github.com/mguterl/chai-datetime/commits)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

jniles committed:
feat(stock): implement getAMC() SP

Implements the getAMC() stored procedure.  It has a different API from
the getCMM() stored procedure that should provide more information and
with better names.

dependabot-preview[bot] committed:
chore(deps-dev): bump karma from 6.1.0 to 6.1.1

Bumps [karma](https://github.com/karma-runner/karma) from 6.1.0 to 6.1.1.
- [Release notes](https://github.com/karma-runner/karma/releases)
- [Changelog](https://github.com/karma-runner/karma/blob/master/CHANGELOG.md)
- [Commits](https://github.com/karma-runner/karma/compare/v6.1.0...v6.1.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps-dev): bump release-it from 14.3.0 to 14.4.0

Bumps [release-it](https://github.com/release-it/release-it) from 14.3.0 to 14.4.0.
- [Release notes](https://github.com/release-it/release-it/releases)
- [Changelog](https://github.com/release-it/release-it/blob/master/CHANGELOG.md)
- [Commits](https://github.com/release-it/release-it/compare/14.3.0...14.4.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps-dev): bump eslint from 7.19.0 to 7.20.0

Bumps [eslint](https://github.com/eslint/eslint) from 7.19.0 to 7.20.0.
- [Release notes](https://github.com/eslint/eslint/releases)
- [Changelog](https://github.com/eslint/eslint/blob/master/CHANGELOG.md)
- [Commits](https://github.com/eslint/eslint/compare/v7.19.0...v7.20.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

jniles committed:
chore: finishing touches on GetAMC()

Migrates the server controllers to using `GetAMC()` instead of
`getCMM()` and updates the test scripts to build the
stock_movement_status table automatically.

jniles committed:
fix: update tests to results from GetAMC().

Updates the stock migration tests to the new output from `GetAMC()`.

jniles committed:
chore: fix AMC modal to reflect API changes

This commit fixes the AMC modal to better reflect the API changes made
with the GetAMC function.

jniles committed:
fix: enable MySQL8 testing again

Attempts to thwart the query optimizer and restore testing on MySQL8.

jmcameron committed:
Fixed adding candidate lots for purchases and transfers

dependabot-preview[bot] committed:
chore(deps): bump p-retry from 4.3.0 to 4.4.0

Bumps [p-retry](https://github.com/sindresorhus/p-retry) from 4.3.0 to 4.4.0.
- [Release notes](https://github.com/sindresorhus/p-retry/releases)
- [Commits](https://github.com/sindresorhus/p-retry/compare/v4.3.0...v4.4.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

jniles committed:
fix(stock): adapt ComputeAMCForStagedInventory()

This commit adapts the ComputeAMCForStagedInventory() stored procedure
to MySQL 5.7.  Hopefully this will be the last change.

jmcameron committed:
Fix lot selection problem

jmcameron committed:
Improve URL path for lot candidates

jniles committed:
fix: do not use decimals for momentjs months

Fixes the display of the days of stock remaining due to momentjs no
longer accepting decimal dates.

jmcameron reviewed (commented):
I did not review the code changes.
Both test:integration and test:integration:stock ran fine on my laptop (WSL Ubuntu 20.04, MySQL 8).
I installed the IMCK db, AND migration files.
I ran CALL zRecomputeStockMovementStatus(); it took 4:40 on my laptop.
The CMM calculations in 'Articles in Stock' seem to be correct.  I spot-checked all 4 algorithms and the calculations were correct.  However Alg 2 still seems to produce very different (higher) estimates than the other 3 algorithms.

Based on direct feedback from @JNiles, I think there will be another review of this in the future; I'll review the code then.

jmcameron committed:
Update the movement creation functions to create lots only if necessary

jmcameron committed:
Fixed code goof

jmcameron committed:
Fix integration test

jniles committed:
fix(stock): add WHERE to max values

Adds a missing WHERE to the max values to appropriately compute the
max_values table calculation.  Seems to also speed up the computation
greatly.

jniles committed:
fix(modal): ensure AMC modal binding is correct

Ensures the AMC/CMM modal `controllerAs` binding is correct.

jniles committed:
fix(stock): improve FR i18n for CMM modal

Improves the French i18n for the CMM modal based on feedback from
@mbayopanda.

jmcameron commented on client/src/modules/stock/inventories/modals/amc.modal.js:
```diff
@@ -12,32 +12,39 @@ function StockAMCModalController(Stock, Notify, Instance, data, moment, Constant
+        const stockOutDate = moment().add(daysOfStockLeft, 'days').toDate();
+
+        // provide this information to the view.
+        vm.data.stock_out_date = stockOutDate;
```
Is there any reason not to combine this with the line above?

jniles replied:
Nope!  I'll do that.

jmcameron commented on server/controllers/inventory/inventory/core.js:
```diff
@@ -259,7 +259,7 @@ function getItemsMetadataById(uid) {
       ig.unique_item, i.consumable, i.locked, i.stock_min, i.sellable,
       i.stock_max, i.created_at AS timestamp, i.type_id, i.unit_id, i.unit_weight, i.unit_volume,
-      ig.sales_account, i.default_quantity, i.avg_consumption, i.delay, i.purchase_interval,
+      ig.sales_account, i.default_quantity, i.delay, i.purchase_interval,
```
The name of this function reminds me that we need an issue to convert the 'items' terminology to 'articles' throughout.

jmcameron commented on server/controllers/stock/import.js:
```diff
@@ -37,6 +37,7 @@ function downloadTemplate(req, res, next) {
 async function importStock(req, res, next) {
   let queryParams;
 
+  const operationDate = new Date(req.body.date);
```
I like being able to specify the date of the import.  It makes sense because the data could be collected at a date before the operator is able to import it.

jmcameron commented on server/controllers/stock/index.js:
```diff
@@ -132,7 +132,7 @@ async function createStock(req, res, next) {
 
     const isExit = 0;
-    const postingParams = [db.bid(documentUuid), isExit, req.session.project.id, req.session.enterprise.currency_id];
+    const postingParams = [db.bid(documentUuid), isExit, req.session.project.id];
```
Why don't we need the currency_id?

jniles replied:
We already know it.  Internally, we value the stock in the enterprise currency id and there isn't any option to change it.  Even if you purchase stock in a foreign currency, we do the internal conversion so that it shows up correctly on the balance sheet.

There isn't anywhere that a user can change the currency associated with stock that I am aware of.

jmcameron reviewed (changes requested):
I looked at the stock import page and had some questions about the import file syntax.  When I downloaded the sample page I see:
![image](https://user-images.githubusercontent.com/62145/108423429-b0f50400-71ec-11eb-8dc2-f919fa40a8af.png)
I suggest changing the group name and inventory_unit so that they do not have any non-ASCII unicode characters.   Also, should the numbers be quoted?

On the importStock page: Shouldn't the import modal support the currency ID?  It does not seem safe to assume the incoming inventory_unit_price will be whatever is the current default for Bhima.

There is a problem with the wording of description for the calculation for Algorithm 4:
![image](https://user-images.githubusercontent.com/62145/108432515-5f06ab00-71f9-11eb-81b4-6ecb458bdc45.png)
Notice that "difference of the number of months in the period minus the total number of days of stock out in the period" implies you are subtracting days from months -- which is a units no-no.  The calculation is correct, but the description has a problem.

I did a units check on algorithms 1, 3 and 4 and they made sense.  While the units do technically work out for algorithm 2, it still does not make sense because it implies that in the next month there would be stock consumed EVERY day at the same rate as it was for the few days stock was consumed during the previous period. Here is what algorithm 2 should be:

   Nsc = number of stock consumption days
   A = pills per stock consumption day
   B = avg number stock consumption days per month

  AMC = A*B = (23 / Nsc)*(Nsc / 184)*30.5
           = (23/3)*(3/184)*30.5 = 3.81

But since there no way to estimate number of days of stock consumed in the future, the estimate has to be based on the past.  Just using the average over the period simply leads Algorithm 3.

mbayopanda commented on client/src/modules/stock/stock.service.js:
```diff
@@ -244,10 +244,12 @@ function StockModalService(Modal) {
     const params = angular.extend(modalParameters, {
       templateUrl,
       controller,
+      controllerAs : 'ModalCtrl',
```
👍🏽

mbayopanda reviewed (approved):
LGTM 👍🏽

jniles committed:
fix: remove unicode/quotes from stock import csv

Removes the unicode and quote characters from the stock import CSV per
recommendation of @jmcameron.
